feat(menu): add icon support for Submenu in Rust and JS/TS APIs

s00d opened the pull request:
This PR adds support for icons (custom and native) in submenus for both the Rust (`tauri`) and JS/TS (`@tauri-apps/api`) APIs.  
Submenus can now display icons just like regular menu items.

<img width="840" alt="image" src="https://github.com/user-attachments/assets/32e72555-fc1c-4916-a81f-3f150eebbabd" />


- `SubmenuOptions` now accepts an `icon` field.
- The `Submenu` class provides `setIcon` and `setNativeIcon` methods.
- Rust API supports icon and nativeIcon for submenus via builder and methods.
- Usage examples are added to the documentation and demo app.

**Closes:** #11796

---

- [x] Changefile added in `.changes/` (minor:enhance)
- [x] All commits are signed
- [x] `cargo test` and `cargo clippy` pass locally for the new code
- [x] PR is ready for review

s00d committed:
feat(menu): add icon and nativeIcon support for Submenu in tauri and @tauri-apps/api

FabianLars commented:
Thank you for the PR! We just released 2.6 so we'll have to keep the dev branch open for patch changes for a few days/weeks. Just so you can prepare for a bit of a delay here :)

Btw, the rust code comments aren't in english yet and your commits (expect the last 2 merge commits) aren't signed :)

s00d committed:
Merge branch 'dev' into dev

s00d commented:
> Thank you for the PR! We just released 2.6 so we'll have to keep the dev branch open for patch changes for a few days/weeks. Just so you can prepare for a bit of a delay here :)
> 
> Btw, the rust code comments aren't in english yet and your commits (expect the last 2 merge commits) aren't signed :)

Thank you for your feedback!
I've fixed the main issues, updated the Rust code comments to English, signed my commits, and squashed everything into a single commit to keep the history clean.
No problem about the delay — I’m happy to wait as long as needed.
Thanks again, and let me know if there’s anything else I should address.

s00d committed:
Merge branch 'dev' into dev

s00d committed:
Merge branch 'dev' into dev

OrangeDoro reviewed (commented):
Hi! I'm a grad student working on a research project about using large language models to automate code review. Based on your commit 45b1a0dde8d8ea588ca7e993ff8230210bc3e8d9 and the changes in crates/tauri/scripts/bundle.global.js, my tool generated this comment:
1. **Error Handling**: Ensure that all asynchronous functions (like `h`) have proper error handling to avoid unhandled promise rejections.
2. **Input Validation**: Validate inputs in functions like `setText`, `setIcon`, and others that accept user input to prevent vulnerabilities.
3. **Callback Management**: Ensure that the `unregisterCallback` method removes callbacks when they are no longer needed to prevent memory leaks.
4. **Memory Management**: Ensure that references stored in `WeakMap` are not held longer than necessary to avoid memory leaks.
5. **Assertions**: Implement assertions to validate expected states or parameters before executing function logic.
6. **Testing Asynchronous Functions**: Ensure that new asynchronous functions have corresponding unit tests to verify their behavior and error handling.
7. **Dynamic Plugin Registration**: Implement a dynamic approach in the `addPluginListener` function to allow for flexible registration of any plugin listener.
8. **Batching Updates**: Consider batching updates in event handling functions to reduce the number of calls to the underlying plugin.
9. **Variable Declarations**: Replace `var` with `let` or `const` for variable declarations to align with modern JavaScript practices.
10. **Function Naming**: Rename functions like `h` to more descriptive names (e.g., `invokePluginMethod`) to clarify their purpose.
11. **Descriptive Variable Names**: Rename single-letter variables (e.g., `e`, `n`, `t`) to more descriptive names to improve code readability.
12. **Consistent Naming Conventions**: Use a consistent naming convention (preferably camelCase) for functions and variables throughout the codebase.
13. **Logging in Asynchronous Functions**: Add logging statements to capture entry and exit points of asynchronous functions for better observability.
14. **Documentation and Comments**: Add comments and documentation for complex logic to help future maintainers understand the code.
15. **Use of Constants for Event Names**: Define event names as constants in a dedicated configuration file for easier management and updates.
16. **No Newline at End of File**: Add a newline at the end of the file to adhere to common coding conventions.
17. **Separation of Concerns**: Separate concerns into different classes or modules for better maintainability and extensibility.

As part of my research, I'm trying to understand how useful these comments are in real-world development. If you have a moment, I'd be super grateful if you could quickly reply to these two yes/no questions:


1. Does this comment provide suggestions from a dimension you hadn’t considered?

2. Do you find this comment helpful?

Thanks a lot for your time and feedback! And sorry again if this message is a bother.

Legend-Master committed:
Merge remote-tracking branch 'upstream/dev' into s00d/dev

Legend-Master commented:
Didn't look at the details yet, but it doesn't seem to work for me on Windows (clicked `Create menu with Image icon` and then `Popup`), is it supposed to work on Windows?

<img width="1502" height="1246" alt="image" src="https://github.com/user-attachments/assets/17df1b3b-14c6-46c1-8254-9ac22773a100" />

s00d commented:
> Didn't look at the details yet, but it doesn't seem to work for me on Windows (clicked `Create menu with Image icon` and then `Popup`), is it supposed to work on Windows?
> 
> <img alt="image" width="1502" height="1246" src="https://private-user-images.githubusercontent.com/68118705/470391544-17df1b3b-14c6-46c1-8254-9ac22773a100.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM0Mjk2ODgsIm5iZiI6MTc1MzQyOTM4OCwicGF0aCI6Ii82ODExODcwNS80NzAzOTE1NDQtMTdkZjFiM2ItMTRjNi00NmMxLTgyNTQtOWFjMjI3NzNhMTAwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI1VDA3NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiMDBlZTViYzQxODllZjVlNGY4YTkyZmNiOTA1OWRiOWM4ZjhiMjgxNGQ0YjlmM2FjYThkYmE5MmY0MDE0MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bLkoc91NdqqOOsvmcnmdPrZUppQB3ZVlfKQ1cga7qQQ">

I apologize, I overlooked that menu formation works slightly differently on Windows in the tray context, so I've created a pull request with the necessary fixes

[PR #308](https://github.com/tauri-apps/muda/pull/308)

Legend-Master committed:
Merge branch 'dev' into s00d/dev

Legend-Master committed:
Update muda

Legend-Master commented on packages/api/src/menu/submenu.ts:
```diff
@@ -286,4 +294,29 @@ export class Submenu extends MenuItemBase {
+      rid: this.rid,
+      icon
+    })
+  }
```
I believe this is covered by `setIcon` right? the `icon` can already take a `NativeIcon`

s00d replied:
yes you are right

Legend-Master commented on packages/api/src/menu/submenu.ts:
```diff
@@ -46,8 +48,14 @@ export function itemFromKind([rid, id, kind]: [number, string, ItemKind]):
+   * Icon to be used for the submenu.
+   * Note: you may need the `image-ico` or `image-png` Cargo features to use this API.
+   */
+  icon?: NativeIcon | string | Image | Uint8Array | ArrayBuffer | number[]
```
Maybe we can extract this to a type alias now (something like `MenuIcon`) with this being used in multiple places

s00d replied:
I moved the type to packages/api/src/image.ts. If it needs to be relocated elsewhere, please let me know.

Legend-Master reviewed (commented):
Looks really good, and works for me on Windows as well now, left some comments and will do a proper review tomorrow

s00d committed:
feat(menu): add set_icon and set_native_icon methods to set submenu icons

s00d committed:
feat(menu): unify icon handling by introducing MenuIcon type

s00d committed:
chore: sync bundle.global.js

s00d committed:
Merge remote-tracking branch 'origin/dev' into dev

s00d commented:
@Legend-Master Fixed your comments. Also added set_icon and set_native_icon in crates/tauri/src/menu/submenu.rs.

Legend-Master committed:
Make setIcon actually work

Legend-Master committed:
Merge branch 'dev' into s00d/dev

Legend-Master committed:
Regenerate `bundle.global.js`

Legend-Master reviewed (approved):
Awesome work, thanks!
